File: frameworks/payload.core/src/main/java/com/qomolangma/frameworks/payload/core/Payload.java
package com.qomolangma.frameworks.payload.core;


import lombok.Getter;

import java.util.LinkedHashMap;

import static com.qomolangma.frameworks.payload.core.MapFieldType.asMap;

@Getter
public class Payload {
    private final Object value;

    public Payload(Object value) {
        this.value = value;
    }

    public <I> I get(FieldType<I> type) {
        return type.from(value);
    }

    public <I> I get(String key, FieldType<I> type) {
        return type.from(get(asMap()).get(key));
    }

    public static ValueBuilder append(String key, Object value) {
        ValueBuilder builder = new ValueBuilder();
        builder.append(key, value);
        return builder;
    }

    public static class ValueBuilder {
        private final LinkedHashMap<String, Object> value;

        public ValueBuilder() {
            this.value = new LinkedHashMap<>();
        }

        public ValueBuilder append(String key, Object value) {
            this.value.put(key, value);
            return this;
        }

        public Payload build() {
            return new Payload(this.value);
        }
    }
}

File: frameworks/payload.core/src/main/kotlin/com/qomolangma/frameworks/payload/core/Payload.kt
package com.qomolangma.frameworks.payload.core

import com.qomolangma.frameworks.payload.core.MapFieldType.Companion.asMap

class Payload(private val value: Any) {
    operator fun <I> get(type: FieldType<I>): I {
        return type.from(value)
    }

    operator fun <I> get(key: String?, type: FieldType<I>): I {
        return type.from(get(asMap())[key])
    }

    class PayloadBuilder {
        private val value: LinkedHashMap<String, Any> = LinkedHashMap()

        fun append(key: String, value: Any): PayloadBuilder {
            this.value[key] = value
            return this
        }

        fun build(): Payload {
            return Payload(value)
        }
    }

    companion object {
        fun append(key: String, value: Any): PayloadBuilder {
            val builder = PayloadBuilder()
            builder.append(key, value)
            return builder
        }
    }
}
